File: examples/nirfmx/wlangen/csharp/VS2010/80211acGenerateWaveform (MultiUser)/cs/Program.cs
using System;
using NationalInstruments.RFToolkits.Wlan.Generation;
using NationalInstruments.ModularInstruments.NIRfsg;
using NationalInstruments.ModularInstruments.NIRfsg.Internal;
using System.Runtime.InteropServices;


namespace _80211acGenerateWaveformMultiUser
{
   class Program
   {
      static void Main(string[] args)
      {
         GenerateeWaveformMultiUserExample example = new GenerateeWaveformMultiUserExample();
         example.Run();
      }
   }

   class GenerateeWaveformMultiUserExample
   {
      const int maxWlanChannels = 4;
      int numTx = 2;
      int numberOfUsers = 2;

      double carrierFrequency = 5.18e9;
      double channelBandwidth = 20e+6;
      double[] powerLevel = { -10.0, -10.0, -10.0, -10.0 };
      double[] externalAttenuation = { 0, 0, 0, 0 };

      NIRfsg[] rfsgSessions = new NIRfsg[maxWlanChannels];
      niWLANG wlanSession;
      NIRfsg externalLOSession = null;
      String[] rfsgResourceName = { "RIO0", "RIO1", "RIO2", "RIO3" };
      String rfsgClockSource = RfsgFrequencyReferenceSource.PxiClock;

      String waveformName = "Wlan";
      private int[] mCSIndex = new int[] { 0, 8 };
      private int[] numOfSpaceTimeStreams = { 1, 1 };
      int[] triggerLines = { 0, 1 };

      public void Run()
      {
         try
         {
            ConfigureWlanGenerationSession();
            ConfigureRfsgSession();
            CreateAndDownloadWaveform();
            StopGeneration();
         }
         catch (Exception e)
         {
            Console.WriteLine("Error : " + e.ToString());
            Console.WriteLine(" Press any key to exit");
            Console.ReadKey();
         }
         finally
         {
            for (int i = 0; i < numTx; i++)
            {
               if (rfsgSessions[i] != null)
                  rfsgSessions[i].Close();
            }

            if (wlanSession != null)
               wlanSession.Close();
         }
      }

      private void ConfigureWlanGenerationSession()
      {
         int isNewSession;
         int mappingMatrixType = niWLANGConstants.MappingMatrixTypeDirect;

         if (wlanSession == null)
            wlanSession = new niWLANG("WLANG", niWLANGConstants.CompatibilityVersion060000, out isNewSession);

         wlanSession.SetStandard(null, niWLANGConstants.Standard80211AcMimoOfdm);
         wlanSession.SetNumberOfTransmitChannels(null, numTx);
         wlanSession.SetChannelBandwidth(null, channelBandwidth);
         wlanSession.SetMappingMatrixType(null, mappingMatrixType);
         wlanSession.SetPPDUType(null, niWLANGConstants.PpduTypeMuPpdu);
         wlanSession.SetNumberOfUsers(null, numberOfUsers);

         String activeChannel = "user";

         for (int i = 0; i < numTx; i++)
         {
            wlanSession.SetMCSIndex(activeChannel + i, mCSIndex[i]);
            wlanSession.SetNumberOfSpaceTimeStreams(activeChannel + i, numOfSpaceTimeStreams[i]);
         }

         wlanSession.SetRFBlankingEnabled(null, niWLANGConstants.True);
      }


      private void CreateAndDownloadWaveform()
      {
         double iqRate;
         double[] actualHeadRoom = new double[numTx];
         String script = @"script GenerateWlan
				                repeat forever
					                generate Wlan
				                end repeat
			                end script";


         IntPtr[] rfsgHandle = new IntPtr[numTx];
         IntPtr externalLOHandle = new IntPtr();

         for (int i = 0; i < numTx; i++)
            rfsgHandle[i] = rfsgSessions[i].GetInstrumentHandle().DangerousGetHandle();

         if (externalLOSession != null)
            externalLOHandle = externalLOSession.GetInstrumentHandle().DangerousGetHandle();
         wlanSession.ConfigureMultipleDeviceSynchronization(rfsgHandle, numTx, rfsgClockSource, triggerLines, triggerLines.Length);
         wlanSession.RFSGConfigureFrequencySingleLO(rfsgHandle, niWLANGConstants.LOSourceOnboard, externalLOHandle, carrierFrequency, niWLANGConstants.False, niWLANGConstants.False);
         wlanSession.RFSGCreateAndDownloadMIMOWaveforms(rfsgHandle, null, numTx, waveformName);


         for (int i = 0; i < numTx; i++)
         {
            niWLANG.WLANG_RFSGConfigureScript(rfsgSessions[i].GetInstrumentHandle().DangerousGetHandle(), null, script, powerLevel[i]);
         }
         wlanSession.RFSGMultipleDeviceInitiate(rfsgHandle);

         //Check for successful generation
         CheckGeneration();

         wlanSession.GetIqRate(String.Empty, out iqRate);
         Console.WriteLine("IQ Rate : {0}", iqRate);

         string channelString = "segment0/channel";
         for (int i = 0; i < numTx; i++)
         {
            wlanSession.GetActualHeadroom(channelString + i, out actualHeadRoom[i]);
            Console.WriteLine("Actual HeadRoom User: {0}", actualHeadRoom[i]);
         }
         Console.WriteLine("Press any key to exit");
         Console.ReadKey();
      }

      private void ConfigureRfsgSession()
      {
         for (int i = 0; i < numTx; i++)
         {
            if (rfsgSessions[i] == null)
            {
               rfsgSessions[i] = new NIRfsg(rfsgResourceName[i], true, false);
            }

            rfsgSessions[i].RF.PowerLevelType = RfsgRFPowerLevelType.PeakPower;
            rfsgSessions[i].Arb.GenerationMode = RfsgWaveformGenerationMode.ArbitraryWaveform;
            rfsgSessions[i].RF.ExternalGain = -(externalAttenuation[i]);
         }
      }

      private bool CheckGeneration()
      {
         RfsgGenerationStatus status = RfsgGenerationStatus.InProgress;
         for (int i = 0; i < numTx; i++)
         {
            status = rfsgSessions[i].CheckGenerationStatus();
            if (status == RfsgGenerationStatus.Complete)
               break;
         }
         return Convert.ToBoolean(status);
      }

      private void StopGeneration()
      {
         for (int i = 0; i < numTx; i++)
         {
            if (rfsgSessions[i] != null)
            {
               rfsgSessions[i].Abort();
               rfsgSessions[i].RF.OutputEnabled = false;
               rfsgSessions[i].Utility.Commit();
               niWLANG.WLANG_RFSGClearDatabase(rfsgSessions[i].GetInstrumentHandle().DangerousGetHandle(), "", null);
            }
         }
      }
   }
}

File: examples/nirfmx/wlangen/csharp/VS2010/80211acGenerateWaveform (MultiUser)/vb/Program.vb

Imports NationalInstruments.RFToolkits.Wlan.Generation
Imports NationalInstruments.ModularInstruments.NIRfsg
Imports NationalInstruments.ModularInstruments.NIRfsg.Internal
Imports System.Runtime.InteropServices


Class Program
   Friend Shared Sub Main(args As String())
      Dim example As New GenerateeWaveformMultiUserExample()
      example.Run()
   End Sub
End Class

Class GenerateeWaveformMultiUserExample

   Const maxWlanChannels As Integer = 4
   Private numTx As Integer = 2
   Private numberOfUsers As Integer = 2

   Private carrierFrequency As Double = 5180000000.0
   Private channelBandwidth As Double = 20000000.0
   Private powerLevel As Double() = {-10.0, -10.0, -10.0, -10.0}
   Private externalAttenuation As Double() = {0, 0, 0, 0}

   Private rfsgSessions As NIRfsg() = New NIRfsg(maxWlanChannels - 1) {}
   Private wlanSession As niWLANG
   Private externalLOSession As NIRfsg = Nothing
   Private rfsgResourceName As [String]() = {"RIO0", "RIO1", "RIO2", "RIO3"}
   Private rfsgClockSource As [String] = RfsgFrequencyReferenceSource.PxiClock

   Private waveformName As [String] = "Wlan"
   Private mCSIndex As Integer() = New Integer() {0, 8}
   Private numOfSpaceTimeStreams As Integer() = {1, 1}
   Private triggerLines As Integer() = {0, 1}

   Public Sub Run()
      Try
         ConfigureWlanGenerationSession()
         ConfigureRfsgSession()
         CreateAndDownloadWaveform()
         StopGeneration()
      Catch e As Exception
         Console.WriteLine("Error : " & e.ToString())
         Console.WriteLine(" Press any key to exit")
         Console.ReadKey()
      Finally
         For i As Integer = 0 To numTx - 1
            If rfsgSessions(i) IsNot Nothing Then
               rfsgSessions(i).Close()
            End If
         Next

         If wlanSession IsNot Nothing Then
            wlanSession.Close()
         End If
      End Try
   End Sub

   Private Sub ConfigureWlanGenerationSession()
      Dim isNewSession As Integer
      Dim mappingMatrixType As Integer = niWLANGConstants.MappingMatrixTypeDirect

      If wlanSession Is Nothing Then
         wlanSession = New niWLANG("WLANG", niWLANGConstants.CompatibilityVersion060000, isNewSession)
      End If

      wlanSession.SetStandard(Nothing, niWLANGConstants.Standard80211AcMimoOfdm)
      wlanSession.SetNumberOfTransmitChannels(Nothing, numTx)
      wlanSession.SetChannelBandwidth(Nothing, channelBandwidth)
      wlanSession.SetMappingMatrixType(Nothing, mappingMatrixType)
      wlanSession.SetPPDUType(Nothing, niWLANGConstants.PpduTypeMuPpdu)
      wlanSession.SetNumberOfUsers(Nothing, numberOfUsers)

      Dim activeChannel As [String] = "user"

      For i As Integer = 0 To numTx - 1
         wlanSession.SetScalarAttributeI32(activeChannel & i, niWLANGProperties.McsIndex, mCSIndex(i))
         wlanSession.SetNumberOfSpaceTimeStreams(activeChannel & i, numOfSpaceTimeStreams(i))
      Next

      wlanSession.SetRFBlankingEnabled(Nothing, niWLANGConstants.[True])
   End Sub


   Private Sub CreateAndDownloadWaveform()
      Dim iqRate As Double
      Dim actualHeadRoom As Double() = New Double(numTx - 1) {}
      Dim script As [String] = "script GenerateWlan" & vbCr & vbLf & vbTab & vbTab & vbTab & vbTab & "                repeat forever" & vbCr & vbLf & vbTab & vbTab & vbTab & vbTab & vbTab & "                generate Wlan" & vbCr & vbLf & vbTab & vbTab & vbTab & vbTab & "                end repeat" & vbCr & vbLf & vbTab & vbTab & vbTab & "                end script"

      Dim rfsgHandle As IntPtr() = New IntPtr(numTx - 1) {}
      Dim externalLOHandle As New IntPtr()

      For i As Integer = 0 To numTx - 1
         rfsgHandle(i) = rfsgSessions(i).GetInstrumentHandle().DangerousGetHandle()
      Next

      If externalLOSession IsNot Nothing Then
         externalLOHandle = externalLOSession.GetInstrumentHandle().DangerousGetHandle()
      End If
      wlanSession.ConfigureMultipleDeviceSynchronization(rfsgHandle, numTx, rfsgClockSource, triggerLines, triggerLines.Length)
      wlanSession.RFSGConfigureFrequencySingleLO(rfsgHandle, niWLANGConstants.LOSourceOnboard, externalLOHandle, carrierFrequency, niWLANGConstants.[False], niWLANGConstants.[False])
      wlanSession.RFSGCreateAndDownloadMIMOWaveforms(rfsgHandle, Nothing, numTx, waveformName)


      For i As Integer = 0 To numTx - 1
         niWLANG.WLANG_RFSGConfigureScript(rfsgSessions(i).GetInstrumentHandle().DangerousGetHandle(), Nothing, script, powerLevel(i))
      Next
      wlanSession.RFSGMultipleDeviceInitiate(rfsgHandle)
      'Check for successful generation
      CheckGeneration()

      wlanSession.GetIqRate([String].Empty, iqRate)
      Console.WriteLine("IQ Rate : {0}", iqRate)

      Dim channelString As String = "segment0/channel"
      For i As Integer = 0 To numTx - 1
         wlanSession.GetActualHeadroom(channelString & i, actualHeadRoom(i))
         Console.WriteLine("Actual HeadRoom User: {0}", actualHeadRoom(i))
      Next
      Console.WriteLine("Press any key to exit")
      Console.ReadKey()

   End Sub

   Private Sub ConfigureRfsgSession()
      For i As Integer = 0 To numTx - 1
         If rfsgSessions(i) Is Nothing Then
            rfsgSessions(i) = New NIRfsg(rfsgResourceName(i), True, False)
         End If

         rfsgSessions(i).RF.PowerLevelType = RfsgRFPowerLevelType.PeakPower
         rfsgSessions(i).Arb.GenerationMode = RfsgWaveformGenerationMode.ArbitraryWaveform
         rfsgSessions(i).RF.ExternalGain = -(externalAttenuation(i))
      Next
   End Sub

   Private Function CheckGeneration() As Boolean
      Dim status As RfsgGenerationStatus = RfsgGenerationStatus.InProgress
      For i As Integer = 0 To numTx - 1
         status = rfsgSessions(i).CheckGenerationStatus()
         If status = RfsgGenerationStatus.Complete Then
            Exit For
         End If
      Next
      Return Convert.ToBoolean(status)
   End Function

   Private Sub StopGeneration()
      For i As Integer = 0 To numTx - 1
         If rfsgSessions(i) IsNot Nothing Then
            rfsgSessions(i).Abort()
            rfsgSessions(i).RF.OutputEnabled = False
            rfsgSessions(i).Utility.Commit()
            niWLANG.WLANG_RFSGClearDatabase(rfsgSessions(i).GetInstrumentHandle().DangerousGetHandle(), "", Nothing)
         End If
      Next
   End Sub
End Class
